src/mlpack/methods/reinforcement_learning/replay/random_replay.hpp

```diff
@@ -0,0 +1,165 @@
+/**
+ * @file random_replay.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is an implementation of random experience replay.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_REPLAY_RANDOM_REPLAY_HPP
+#define MLPACK_METHODS_RL_REPLAY_RANDOM_REPLAY_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of random experience replay.
+ *
+ * At each time step, interactions between the agent and the
+ * environment will be saved to a memory buffer. When necessary,
+ * we can simply sample previous experiences from the buffer to
+ * train the agent. Typically this would be a random sample and
+ * the memory will be a First-In-First-Out buffer.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @phdthesis {lin1993reinforcement,
+ *  title  =  {Reinforcement learning for robots using neural networks},
+ *  author =  {Lin, Long-Ji},
+ *  year   =  {1993},
+ *  school =  {Fujitsu Laboratories Ltd}
+ * }
+ * @endcode
+ *
+ * @tparam EnvironmentType Desired task.
+ */
+template <typename EnvironmentType>
+class RandomReplay
+{
+ public:
+  using ActionType = typename EnvironmentType::Action;
+  using StateType = typename EnvironmentType::State;
+
+  /**
+   * Construct an instance of random experience replay class.
+   *
+   * @param batchSize Number of examples returned at each sample.
+   * @param capacity Total memory size in terms of number of examples.
+   * @param dimension The dimension of an encoded state.
+   */
+  RandomReplay(const size_t batchSize,
+               const size_t capacity,
+               const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity(capacity),
+      position(0),
+      states(dimension, capacity),
+      actions(capacity),
+      rewards(capacity),
+      nextStates(dimension, capacity),
+      isTerminal(capacity),
+      full(false)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Store the given experience.
+   *
+   * @param state Given state.
+   * @param action Given action.
+   * @param reward Given reward.
+   * @param nextState Given next state.
+   * @param isEnd Whether next state is terminal state.
+   */
+  void Store(const StateType& state,
+             ActionType action,
+             double reward,
+             const StateType& nextState,
+             bool isEnd)
+  {
+    states.col(position) = state.Encode();
+    actions(position) = action;
+    rewards(position) = reward;
+    nextStates.col(position) = nextState.Encode();
+    isTerminal(position) = isEnd;
+    position++;
+    if (position == capacity)
```

Comment:
Shouldn't we use ``if ((position % capacity) == 0)``here, we only reset position once, so we could run into a index out of bounds error for the second run, maybe I missed something?